tools/pyboard.py: ProcessToSerial cross-platform fix

vikmal-sw opened the pull request:
Using ProcessToSerial class from pyboard.py on Windows was not possible due to some Unix-only code (for example, select.poll or selectors.select()). A solution to get ProcessToSerial working on Windows is by creating a thread that constantly reads from the subp.stdout and completes a queue. Then, the read() method will get the characters from the queue. This solution also works on Unix platforms, but I have implemented it only for Windows platforms, leaving the original code for Unix platforms.

vikmal-sw committed:
tools/pyboard.py: ProcessToSerial cross-platform fix

Using ProcessToSerial class from pyboard.py on Windows was not possible due to some Unix-only code (for example, select.poll or selectors.select()). A solution to get ProcessToSerial working on Windows is by creating a thread that constantly reads from the subp.stdout and completes a queue. Then, the read() method will get the characters from the queue. This solution also works on Unix platforms, but I have implemented it only for Windows platforms, leaving the original code for Unix platforms.

dpgeorge commented on tools/pyboard.py:
```diff
@@ -153,57 +153,114 @@ def inWaiting(self):
+            q.put(data)
+        data_prec = data
 
 class ProcessToSerial:
```
I think it'd be much cleaner to create separate classes, such as: `ProcessToSerialPosix` and `ProcessToSerialThreaded`.  Then some code like:
```python
import select
if hasattr(select, "poll"):
    ProcessToSerial = ProcessToSerialPosix
else:
    ProcessToSerial = ProcessToSerialThreading
```

dpgeorge commented:
Thanks for the contribution!

It's also possible to use `socket.socketpair()` on Windows (and Linux) and use that instead of a `Queue` to communicate between the threads.  The benefit of a socket pair is that you can use `select.select(...)` on it to efficiently wait for data.

vikmal-sw committed:
tools/pyboard.py: Creating 2 separate classes for ProcessToSerial

In order to make the code cleaner, there are 2 separate classes instead of ProcessToSerial :
- ProcessToSerialPosix which uses select module to check if any data available on the subprocess stdout (working well on Unix)
- ProcessToSerialThreading which uses a thread that completes a queue with data from the subprocess stdout (working well on Windows as an alternative to select)
The choice is made by verifying if select has the attribute "poll" on the host platform

vikmal-sw commented:
Thank you very much for your review and answer dpgeorge !
I have created 2 separate classes as you suggested : ProcessToSerialPosix and ProcessToSerialThreading
I have also investigated the possibility of using socket.socketpair(). It is a great idea, but I found that it is supported on Windows only by Python version 3.5 or higher which I am afraid introduces Python 3.5 dependency. What do you think ?
Thank you very much in advance !
Best regards,
Victor Malai

dpgeorge commented on tools/pyboard.py:
```diff
@@ -258,7 +316,12 @@ def __init__(
+            if hasattr(select, "poll"):
+                self.serial = ProcessToSerialPosix(device[len("exec:") :])
+            else:
+                self.serial = ProcessToSerialThreading(device[len("exec:") :])
```
The `ProcessToSerial` class is really part of the public API of this file, so I suggest to put this code in the global scope, so that `ProcessToSerial` is defined in the global scope.  Then the code here doesn't need to change at all.

vikmal-sw replied:
Excuse me, I don't have much experience with Python. If I keep the ProcessToSerial class, how and where I could declare ProcessToSerialPosix and ProcessToSerialThreading ? I appreciate very much your help and indications.
Thank you in advance.
Best regards,
Victor

dpgeorge replied:
```python
class ProcessToSerialPosix:
    ...

class ProcessToSerialThreading:
    ...

if hasattr(select, "poll"):
    ProcessToSerial = ProcessToSerialPosix
else:
    ProcessToSerial = ProcessToSerialThreading
```

vikmal-sw replied:
Thank you very much for your time and help ! I didn't know you can do that in Python. I made a commit with these changes.

dpgeorge commented:
> I have also investigated the possibility of using socket.socketpair(). It is a great idea, but I found that it is supported on Windows only by Python version 3.5 or higher which I am afraid introduces Python 3.5 dependency. What do you think ?

OK, in that case let's stick with Queue.

dpgeorge commented on tools/pyboard.py:
```diff
@@ -153,13 +153,72 @@ def inWaiting(self):
             return n_waiting
 
+# A thread that constantly reads from subp stdout and puts data into a queue (FIFO list)
+# When tested, it reads double '\r' which messes up the further code, so if double '\r' detected, skip one
```
Do you understand why it reads double \r?

vikmal-sw replied:
I have tried to investigate the reason, but unfortunately I am still not able to understand why it reads double \r

vikmal-sw committed:
tools/pyboard.py: Keep ProcessToSerial class in global scope

Considering ProcessToSerial is a part of the public API of this file, it is defined in the global scope. This leaves the Pyboard class code unchanged.

dpgeorge commented:
I'll close this PR due to inactivity.

Nowadays `mpremote` is best to use for control of a board over serial, so any improvements so be made to that.
